calculator.go

```diff
@@ -60,45 +60,34 @@ var functions = map[string]interface{}{
 func call(funcName string, args []float64) (float64, error) {
 	f, ok := functions[funcName]
 	if !ok {
-		return 0, fmt.Errorf("unknown function %s", funcName)
+		return 0, fmt.Errorf("unknown function: %s", funcName)
 	}
-	switch f := f.(type) {
-	case func() float64:
-		return f(), nil
-	case func(float64) float64:
-		return f(args[0]), nil
-	case func(float64, float64) float64:
-		return f(args[0], args[1]), nil
-	case func(float64, float64, float64) float64:
-		return f(args[0], args[1], args[2]), nil
-	default:
-		return 0, fmt.Errorf("invalid function %s", funcName)
-	}
-}
 
-func calculate(n *node) (float64, error) {
-	switch n.kind {
-	case addNode:
-		left, err := calculate(n.left)
-		if err != nil {
-			return 0, err
+	switch len(args) {
+	case 0:
+		if fn, ok := f.(func() float64); ok {
+			return fn(), nil
 		}
-		right, err := calculate(n.right)
-		if err != nil {
-			return 0, err
+	case 1:
+		if fn, ok := f.(func(float64) float64); ok {
+			return fn(args[0]), nil
 		}
-		return left + right, nil
-	case subNode:
-		left, err := calculate(n.left)
-		if err != nil {
-			return 0, err
+	case 2:
+		if fn, ok := f.(func(float64, float64) float64); ok {
+			return fn(args[0], args[1]), nil
 		}
-		right, err := calculate(n.right)
-		if err != nil {
-			return 0, err
+	case 3:
+		if fn, ok := f.(func(float64, float64, float64) float64); ok {
+			return fn(args[0], args[1], args[2]), nil
 		}
-		return left - right, nil
-	case mulNode:
+	}
+
+	return 0, fmt.Errorf("invalid function signature for %s", funcName)
+}
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The new implementation of the `call` function is more readable and maintainable than the previous one. It checks the length of the arguments and then tries to cast the function to the appropriate type based on the number of arguments. However, there's a potential issue with this approach: if the number of arguments does not match any of the cases in the switch statement, an error will be returned. This could lead to unexpected behavior if a function with more than 3 arguments is added to the `functions` map. Consider adding a default case to handle functions with more than 3 arguments.

```diff
+	default:
+		return 0, fmt.Errorf("unsupported number of arguments for %s", funcName)
```



<!-- This is an auto-generated comment by OSS CodeRabbit -->